tempesta_fw/sched/tfw_sched_hash.c

```diff
@@ -175,109 +211,127 @@ tfw_sched_hash_get_sg_conn(TfwMsg *msg, TfwSrvGroup *sg)
+__add_conn(TfwHashConnList *cl, TfwSrvConn *conn, unsigned long hash)
+{
+	ssize_t idx;
+	TfwHashConn new_hcon = {hash, conn};
+
+	idx = __bsearch(hash, cl);
+	if (idx < 0) {
+		/* @cl is empty. */
+		cl->conns[0] = new_hcon;
+		++cl->conn_n;
+
+		return 0;
+	}
+
+	if (cl->conns[idx].hash == hash)
+		return -EEXIST;
+
+	/* Need to insert connection before or after found index. */
+	if ((hash > cl->conns[idx].hash) && (idx != (ssize_t)cl->conn_n))
+		++idx;
+
+	if ((size_t)idx != cl->conn_n)
+		memmove(&cl->conns[idx+1], &cl->conns[idx],
+			sizeof(TfwHashConn) * (cl->conn_n - (size_t)idx));
+
+	cl->conns[idx] = new_hcon;
+	++cl->conn_n;
+
+	return 0;
+}
+
+static void
+__fill_srv_lists(TfwHashConnList *cl)
+{
+	size_t i;
 
-	if (!sl)
-		return;
+	for (i = 0; i < cl->conn_n; ++i) {
+		TfwHashConn *hconn = &cl->conns[i];
+		TfwServer *srv = (TfwServer *)hconn->conn->peer;
+		TfwHashConnList *scl = srv->sched_data;
 
-	for (si = 0; si < sl->srv_n; ++si)
-		if (sl->srvs[si].conn)
-			kfree(sl->srvs[si].conn);
-	kfree(sl);
-	sg->sched_data = NULL;
+		scl->conns[scl->conn_n] = *hconn;
+		++scl->conn_n;
+	}
 }
 
 static int
 tfw_sched_hash_add_grp(TfwSrvGroup *sg)
 {
-	int ret = -EINVAL;
-	size_t size, si, ci;
-	unsigned int sum_conn_n;
+	int ret;
+	size_t size, conn_n = 0, seed = 0;
 	TfwServer *srv;
-	TfwHashSrv *hsrv;
-	TfwHashSrvList *sl;
+	TfwHashConnList *cl;
 
 	if (unlikely(!sg->srv_n || list_empty(&sg->srv_list)))
 		return -EINVAL;
 
-	size = sizeof(TfwHashSrvList) + sizeof(TfwHashSrv) * sg->srv_n;
+	list_for_each_entry(srv, &sg->srv_list, list) {
+		conn_n += srv->conn_n;
+	}
+
+	size = sizeof(TfwHashConnList) + sizeof(TfwHashConn) * conn_n;
 	if (!(sg->sched_data = kzalloc(size, GFP_KERNEL)))
 		return -ENOMEM;
-	sl = sg->sched_data;
-	sl->srvs = sg->sched_data + sizeof(TfwHashSrvList);
-	sl->srv_n = sg->srv_n;
+	cl = sg->sched_data;
 
-	si = sum_conn_n = 0;
-	hsrv = sl->srvs;
 	list_for_each_entry(srv, &sg->srv_list, list) {
-		TfwSrvConn **conn, *srv_conn;
-		unsigned long *hash;
-
-		if (unlikely((si++ == sg->srv_n) || !srv->conn_n
-			     || list_empty(&srv->conn_list)))
-			goto cleanup;
-
-		size = (sizeof(hsrv->conn[0]) + sizeof(hsrv->hash[0]))
-		       * srv->conn_n;
-		if (!(hsrv->conn = kzalloc(size, GFP_KERNEL))) {
-			ret = -ENOMEM;
-			goto cleanup;
-		}
-		hsrv->hash = (typeof(hsrv->hash))(hsrv->conn + srv->conn_n);
-
-		ci = 0;
-		conn = hsrv->conn;
-		hash = hsrv->hash;
-		list_for_each_entry(srv_conn, &srv->conn_list, list) {
-			if (unlikely(ci++ == srv->conn_n))
+		TfwSrvConn *conn;
+		unsigned long srv_hash = __calc_srv_hash(srv);
+
+		size = sizeof(TfwHashConnList)
+				+ sizeof(TfwHashConn) * srv->conn_n;
+		if (!(srv->sched_data = kzalloc(size, GFP_KERNEL)))
+			return -ENOMEM;
+
+		list_for_each_entry(conn, &srv->conn_list, list) {
+			unsigned long hash;
+retry:
+			seed += 1;
+			hash = hash_long(srv_hash ^ seed, BITS_PER_LONG);
+
+			ret = __add_conn(cl, conn, hash);
+			if (ret) {
+				if (ret == -EEXIST)
```

Comment:
Only `-EEXIST` is returned from `__add_conn()` as error code, so no need for `ret` variable.